### PR TITLE
feat(cpn): support more than just FrSky S.Port telemetry simulation

### LINKED ISSUE #2156: ExpressLRS telemetry on simulator
### Companion Simulator Telemetry
Now the expressLRS is more and more integrative in edgeTx,
is is very hard to debug lua script that support expressLRS 

### suggested solution

enhance the "telemetry" screen on the companion simulator with the elrs default telemetry
that will be easier to build and debug lua script that support expressLRS as well

add the following:
image

add them to here:
image

### Describe alternatives you've considered

_No response_

### Additional context

_No response_

### LINKED ISSUE #4983: Companion Telemetry Simulator does not support ELRS/CRSF Telemetry
### Is there an existing issue for this problem?

- [X] I have searched the existing issues

### What part of EdgeTX is the focus of this bug?

Simulator

### Current Behavior

I am using a Radiomaster 2.4 GHz ELRS-Module and a Radiomaster ER8GV Receiver having a Vario).
I was not able, to simulate telemetry values for this configuration.
Screenshots of the Telemetry page of the Transmitter and the Telemetry Simulator screen are  attached in this report. It seems, that the Telemtry Simulator is limited to FrSky Telemetry values. ELRS uses the Crossfire protocol and all telemetry values appear as "Custom" on the Radio.
Telemetry-Screen
Telementry-Simulator

### Expected Behavior

I would like to have possibility, to configure custom sensors for the Telemetry Simulator or some kind od automatic recognition of the telemetry values.

### Steps To Reproduce

- Configure a radio for an ELRS receiver
- Discover all sensors
- Save configuration to Companion
- Start a Simulation
- Modify Values (e.g. Vario Alt)
Sensors do not match the ones in the radio

### Version

2.9.4

### Transmitter

FrSky X12

### Operating System (OS)

Linux

### OS Version

LMDE 6 Faye

### Anything else?

_No response_

### PR DESCRIPTION
**Pull request by nrw505**
Fixes #4983
Fixes #2156 

Summary of changes:

1. Replace the current telemetry sensor UI dropdowns to choose a telemetry source for both the internal and external module.
2. Refactor the current FrSky S.port telemetry implementation to allow it to have separate telemetry for both internal and external modules at the same time
3. Update the simulator interface to accept different telemetry protocols
4. Implement a CRSF telemetry source and associated UI
5. Rejigger the way the log replay controller works so that it can work with multiple kinds of telemetry

Example: [Screencast from 10-08-24 01:17:55.webm](https://github.com/user-attachments/assets/fadfb3a1-dae1-4305-aa94-da30b1b10c8d)

### CODE REVIEW

**Comment by elecpower** [comment]
Post merge of #5381 sim telemetry will need to be modified to suit surface radios eg flight vs drive and hide flight sensors eg vario

File: companion/src/simulation/simulatedgps.h [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ */
+
+#ifndef _SIMULATEDGPS_H_
+#define _SIMULATEDGPS_H_
```
Comment (elecpower) [inline companion/src/simulation/simulatedgps.h:23]: Please use #pragma once for header guards throughout. It is a requirement for all new code and is progressively being performed as housekeeping by the dev team unless it does not achieve the desired compile code.

File: companion/src/simulation/telemetryprovider.h [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ */
+
+#ifndef _TELEMETRYPROVIDER_H_
+#define _TELEMETRYPROVIDER_H_
```
Comment (elecpower) [inline companion/src/simulation/telemetryprovider.h:23]: Fix required

File: companion/src/simulation/telemetryprovidercrossfire.cpp [hunk 3]
```diff
@@ -0,0 +1,530 @@
...
+  connect(ui->input_alt,  SIGNAL(valueChanged(double)),
+          &gps,           SLOT(setAltitude(double)));
+  connect(ui->input_sats, SIGNAL(valueChanged(int)),
+          &gps,           SLOT(setSatelliteCount(int)));
+
+  connect(&gps,           &SimulatedGPS::positionChanged,
+          ui->input_gps,  &QLineEdit::setText);
+  connect(&gps,           &SimulatedGPS::courseDegreesChanged,
+          [this](double v) { ui->input_hdg->setValue(v); });
+
```
Comment (elecpower) [inline companion/src/simulation/telemetryprovidercrossfire.cpp:60]: For code consistency with existing code please keep Qt::connect functions to a single line. We usually only split across lines when lambda functions used for the receiver.
Please replace the use of SIGNAL and SLOT macros with function pointers and lambda functions. The macros do not provide compile time checking of the sender and receiver events.

File: companion/src/simulation/telemetryprovidercrossfire.h [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ */
+
+#ifndef _TELEMETRY_CROSSFIRE_H_
+#define _TELEMETRY_CROSSFIRE_H_
```
Comment (elecpower) [inline companion/src/simulation/telemetryprovidercrossfire.h:23]: Amend

File: companion/src/simulation/telemetryproviderfrsky.h [hunk 5]
```diff
@@ -0,0 +1,114 @@
...
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ */
+
+#ifndef _TELEMETRY_FRSKY_H_
+#define _TELEMETRY_FRSKY_H_
```
Comment (elecpower) [inline companion/src/simulation/telemetryproviderfrsky.h:23]: Amend

File: companion/src/simulation/telemetrysimu.cpp [hunk 6]
```diff
@@ -76,7 +67,16 @@ TelemetrySimulator::TelemetrySimulator(QWidget * parent, SimulatorInterface * si
...
+  connect(ui->internalTelemetrySelector, &QComboBox::currentTextChanged,
+          this,                          &TelemetrySimulator::onInternalTelemetrySelectorChanged);
+  connect(ui->externalTelemetrySelector, &QComboBox::currentTextChanged,
+          this,                          &TelemetrySimulator::onExternalTelemetrySelectorChanged);
+
+  connect(this,                &TelemetrySimulator::internalTelemetryDataChanged,
+          simulator,           &SimulatorInterface::sendInternalModuleTelemetry);
+  connect(this,                &TelemetrySimulator::externalTelemetryDataChanged,
+          simulator,           &SimulatorInterface::sendExternalModuleTelemetry);
+
```
Comment (elecpower) [inline companion/src/simulation/telemetrysimu.cpp:79]: Amend

File: companion/src/simulation/telemetrysimu.cpp [hunk 7]
```diff
@@ -153,53 +161,78 @@ void TelemetrySimulator::onSimulateToggled(bool isChecked)
     stopTelemetry();
 }
 
-void TelemetrySimulator::onLogTimerEvent()
+TelemetryProvider * TelemetrySimulator::newTelemetryProviderFromDropdownChoice(const QString &text, QScrollArea * parent, bool isExternal)
 {
-  logPlayback->stepForward(false);
+  if (text == "None") {
```
Comment (elecpower) [inline companion/src/simulation/telemetrysimu.cpp:166]: This an subsequent branch tests will not achieve the desired result when the text is translated in the drop down list. Use the currentIndex instead

File: companion/src/simulation/telemetrysimu.cpp [hunk 8]
```diff
@@ -153,53 +161,78 @@ void TelemetrySimulator::onSimulateToggled(bool isChecked)
...
+    return NULL;
+  }
+  if (text == "FrSky S.Port") {
```
Comment (elecpower) [inline companion/src/simulation/telemetrysimu.cpp:169]: amend

File: companion/src/simulation/telemetrysimu.cpp [hunk 9]
```diff
@@ -153,53 +161,78 @@ void TelemetrySimulator::onSimulateToggled(bool isChecked)
...
+    if (isExternal) {
+      connect(newProvider, &TelemetryProviderFrSky::telemetryDataChanged,
+              this,        &TelemetrySimulator::onExternalTelemetryProviderDataChanged);
+    } else {
+      connect(newProvider, &TelemetryProviderFrSky::telemetryDataChanged,
+              this,        &TelemetrySimulator::onInternalTelemetryProviderDataChanged);
+    }
+    return newProvider;
+  }
+  if (text == "CRSF / ELRS") {
```
Comment (elecpower) [inline companion/src/simulation/telemetrysimu.cpp:181]: amend

File: companion/src/simulation/telemetrysimu.cpp [hunk 10]
```diff
@@ -153,53 +161,78 @@ void TelemetrySimulator::onSimulateToggled(bool isChecked)
...
+  }
+  if (text == "FrSky S.Port") {
+    TelemetryProviderFrSky * newProvider =  new TelemetryProviderFrSky(parent);
+    parent->setWidget(newProvider);
+    if (isExternal) {
+      connect(newProvider, &TelemetryProviderFrSky::telemetryDataChanged,
+              this,        &TelemetrySimulator::onExternalTelemetryProviderDataChanged);
+    } else {
+      connect(newProvider, &TelemetryProviderFrSky::telemetryDataChanged,
+              this,        &TelemetrySimulator::onInternalTelemetryProviderDataChanged);
```
Comment (elecpower) [inline companion/src/simulation/telemetrysimu.cpp:177]: amend connects

File: companion/src/simulation/telemetrysimu.cpp [hunk 11]
```diff
@@ -153,53 +161,78 @@ void TelemetrySimulator::onSimulateToggled(bool isChecked)
...
+  }
+  if (text == "CRSF / ELRS") {
+    TelemetryProviderCrossfire * newProvider =  new TelemetryProviderCrossfire(parent);
+    parent->setWidget(newProvider);
+    if (isExternal) {
+      connect(newProvider, &TelemetryProviderCrossfire::telemetryDataChanged,
+              this,        &TelemetrySimulator::onExternalTelemetryProviderDataChanged);
+    } else {
+      connect(newProvider, &TelemetryProviderCrossfire::telemetryDataChanged,
+              this,        &TelemetrySimulator::onInternalTelemetryProviderDataChanged);
```
Comment (elecpower) [inline companion/src/simulation/telemetrysimu.cpp:189]: amend

File: companion/src/simulation/telemetrysimu.h [hunk 12]
```diff
@@ -28,8 +28,10 @@
 #include <QDateTime>
 #include <QtCore/qmath.h>
 #include <QFileDialog>
+#include <QScrollArea>
```
Comment (elecpower) [inline companion/src/simulation/telemetrysimu.h:31]: Header guard above

File: companion/src/simulation/telemetryprovidercrossfire.cpp [hunk 13]
```diff
@@ -0,0 +1,530 @@
...
+    QMessageBox::critical(this, CPN_STR_APP_NAME, tr("Unable to open file for reading.\n%1").arg(file.errorString()));
+    return;
+  }
+
+  QTextStream in(&file);
+
+  QString inputText;
+  double inputDouble;
+  int inputInt;
+
```
Comment (elecpower) [inline companion/src/simulation/telemetryprovidercrossfire.cpp:495]: Currently possible to inadvertently load Frsky saved values. Needs a test and rejection msg to save the user from themselves

File: companion/src/simulation/telemetryproviderfrsky.cpp [hunk 14]
```diff
@@ -0,0 +1,934 @@
...
+        return;
+
+    QFile file(idFileNameAndPath);
+
+    if (!file.open(QIODevice::ReadOnly)){
+        QMessageBox::critical(this, CPN_STR_APP_NAME, tr("Unable to open file for reading.\n%1").arg(file.errorString()));
+        return;
+    }
+
+    QTextStream in(&file);
```
Comment (elecpower) [inline companion/src/simulation/telemetryproviderfrsky.cpp:776]: Currently possible to inadvertently load crsf saved values. Needs a test and rejection msg to save the user from themselves

**Review by elecpower — CHANGES_REQUESTED** [review]
UI comments:
- change buttons font size to default (just look out of place in Linux)
- QComboBox protocol selector would look better IMO if moved closer to label

**Comment by nrw505** [comment]
> * change buttons font size to default (just look out of place in Linux)

They already are the default font size, the .ui file does not specify a font size for the buttons e.g.

```
    <widget class="QPushButton" name="button_saveTelemetryValues">
     <property name="text">
      <string>Save Telemetry Values</string>
     </property>
    </widget>
```

This is the same font size that the buttons are in the current frsky-only telemetry screen (and the fields and labels are 8pt as they are on the current frsky-only telemetry screen).

I'm happy to change them to 8 point so they match the rest of the fields on that widget it that's what you want. Please confirm.

> * QComboBox protocol selector would look better IMO if moved closer to label

Just right-align the labels so they're next to the QComboBox OK?

**Comment by elecpower** [comment]
> This is the same font size that the buttons are in the current frsky-only telemetry screen (and the fields and labels are 8pt as they are on the current frsky-only telemetry screen).
> I'm happy to change them to 8 point so they match the rest of the fields on that widget it that's what you want. Please confirm.

Make buttons consistent with rest of dialog.

> Just right-align the labels so they're next to the QComboBox OK?

Right align would suffice.

IMO the dialog could do with some grids to aid in layout but that is for another day.

**Review by elecpower — DISMISSED** [review]

**Comment by LupusTheCanine** [comment]
whould it be possible to use serial port as telemetry source and possibly output port for protocols like CRSF?

**Comment by nrw505** [comment]
> whould it be possible to use serial port as telemetry source and possibly output port for protocols like CRSF?

If you wanted to do that I think you'd probably be better off changing radio/src/targets/simu to use a real serial port as an internal (or external) module - then you'd have _real_ telemetry and the simulation telemetry wouldn't be used at all.

**Comment by pfeerick** [comment]
My only nitpicks so far would be:
- for CRSF/ELRS, FM has quotes which shouldn't be there
- 1RSS isn't being read from the attached log file (TX16S, ELRS, Betaflight).  [TinyTrainer2-2024-08-03-091029.zip](https://github.com/user-attachments/files/16644282/TinyTrainer2-2024-08-03-091029.zip)
- Could do with some padding at the top for the checkboxes... 
image

Otherwise, after a quick poke around and replay of the log... this is working fantastic! :)

**Comment by pfeerick** [comment]
btw, don't think I didn't notice this! 🤣 🤣 🤣 
image

**Comment by nrw505** [comment]
> * for CRSF/ELRS, FM has quotes which shouldn't be there

Fixed

> * 1RSS isn't being read from the attached log file (TX16S, ELRS, Betaflight).  [TinyTrainer2-2024-08-03-091029.zip](https://github.com/user-attachments/files/16644282/TinyTrainer2-2024-08-03-091029.zip)

It's being read OK, but I had the UI set up wrong given CSRF RSS measurements are an actual measurement not a percentage-like value like S.Port. Fixed.

> * Could do with some padding at the top for the checkboxes... (ignore this if already covered earlier)

Oops, didn't mean to change that part of the original UI. Padding added.

**Comment by nrw505** [comment]
> btw, don't think I didn't notice this! 🤣 🤣 🤣 image

Hey, you said you wanted it merge-ready :P . Keep an eye out for followup PRs. I also want to change what's selectable in the internal/external module dropdowns based on what the simulated radio is capable of and maybe a warning if you choose something that the current model isn't configured for.

**Comment by 3djc** [comment]
Looks quite nice to me. 

Unsure if in the scope of this contest, but I think ELRS has this specificity that you can set frame rate and telemetry frame ratio. When set to very low frame rate (like 25hz) and low telemetry ratio, the flow of incoming telemetry is quite slow, and may potentially affect Lua scripts. It would be great if (even at a later time) you could have a way to specify 'I'm running at 25Hz, 1/32 telem' (or whatever those values are)

But already quite nice !

**Comment by pfeerick** [comment]
That's probably out of scope, but would be interesting to see. Do you see that primarily only affecting the replay of a log? Thus effectively just more granularity over the replay rate? Or are you after it slowing down the refreshes of the sensors, i.e. for ELRS to be taking into account the [LINK and DATA groupings and data rates for sensors](https://www.expresslrs.org/info/telem-bandwidth/).

**Comment by 3djc** [comment]
I was thinking 'live' usage, affecting the refresh of sensors, for LUA devs who might make computation based on sensor that might glitch at low or high refresh speed

**Comment by nrw505** [comment]
Oops wrong button sorry

**Comment by 3djc** [comment]
I was not thinking log at all, only live data, that could have a "jitter" value of 0 (so like today) or say x% where value would vary by that much, as it is currently difficult when you debug scripts on simu to see how your values follow (or not) sensor value changes (but then again , out of scope of the competition itself)

**Comment by nrw505** [comment]
@3djc you mean kind of like how the simulated GPS can currently be set to run its own little simulation which updates the position over time?

**Comment by 3djc** [comment]
> @3djc you mean kind of like how the simulated GPS can currently be set to run its own little simulation which updates the position over time?

Something like that
image

where you could decide an amount of variation. If you wanted to go over the top, you would add another column to decide variation type (rng, sine, triangle, square,..), but a default of triangle or rng would  be perfect

**Comment by nrw505** [comment]
Rebased to squash all the fixups

**Comment by pfeerick** [comment]
Probably my final comments on this before merge:
- can we make it so don't need to re-enter instance ids for the FrSky SPort fields before they function? As now need to reenter "something" before can replay frsky telemetry.
- from a BF ELRS quad, GPS lat long appears to not be comma separated, so doesn't relay through to simulation from the logs (i.e. see [Crux3 NLR-2022-10-08-093345.csv.zip](https://github.com/user-attachments/files/16667282/Crux3.NLR-2022-10-08-093345.csv.zip))...

**Comment by nrw505** [comment]
> * can we make it so don't need to re-enter instance ids for the FrSky SPort fields before they function? As now need to reenter at least "something" before can replay frsky telemetry.

Whoops, had some initialisation for the FrSky provider called by the wrong spot: now it populates those instance id fields like it did before this PR.

> * from a BF ELRS quad, GPS lat long appears to not be comma separated, so doesn't relay through to simulation from the logs

Made the simulated GPS accept space separated lat/long as well as tolerating spaces around a comma.

**Pull request merged**